presto-main/src/test/java/com/facebook/presto/operator/TestHashSemiJoinOperator.java

```diff
@@ -130,6 +144,134 @@ public void testSemiJoin(boolean hashEnabled)
+                assertTrue(setBuilderOperator.getOperatorContext().isWaitingForMemory().isDone());
+
+                // free the pool for the next iteration
+                memoryPool.free(queryId, reservedMemoryInBytes);
+                continue;
+            }
+
+            // assert the hash capacity is not changed; otherwise, we should have yielded because we reduced the memory pool size.
+            long newHashCapacity = (long) ((SetBuilderOperator)setBuilderOperator).getCapacity();
+            assertTrue(oldCapacity == newHashCapacity);
+
+            if (fNeedsInput) {
+                // We have successfully added a page
+
+                // Assert we are not blocked
+                assertTrue(setBuilderOperator.getOperatorContext().isWaitingForMemory().isDone());
+
+                // We are not going to rehash; therefore, assert the memory increase only comes from the aggregator
+                assertLessThan(actualIncreasedMemory, additionalMemoryInBytes);
+
+                // free the pool for the next iteration
+                memoryPool.free(queryId, reservedMemoryInBytes);
+            }
+            else {
+                // We failed to finish the page processing i.e. we yielded
+                yieldCount++;
+
+                // Assert we are blocked
+                assertFalse(setBuilderOperator.getOperatorContext().isWaitingForMemory().isDone());
+
+                // Increased memory is no smaller than the hash table size and no greater than the hash table size + the memory used by aggregator
+                // Note that we only put the first block(column) into the hash table so we only need to count the first blocks retained size.
+                if (type == BIGINT) {
+                    // groupIds and values double by hashCapacity; while valuesByGroupId double by maxFill = hashCapacity / 0.75
+                    expectedReservedExtraBytes = oldCapacity * (long) (Long.BYTES * 1.75 + Integer.BYTES) + page.getBlock(0).getRetainedSizeInBytes();
+                }
+                else {
+                    // groupAddressByHash, groupIdsByHash, and rawHashByHashPosition double by hashCapacity; while groupAddressByGroupId double by maxFill = hashCapacity / 0.75
+                    expectedReservedExtraBytes = oldCapacity * (long) (Long.BYTES * 1.75 + Integer.BYTES + Byte.BYTES) + page.getBlock(0).getRetainedSizeInBytes();
+                }
+                // THIS FAILS NOW actualIncreasedMemory < expectedReservedExtraBytes
+                assertBetweenInclusive(actualIncreasedMemory, expectedReservedExtraBytes, expectedReservedExtraBytes + additionalMemoryInBytes);
+
+                // Free the pool to unblock
+                memoryPool.free(queryId, reservedMemoryInBytes);
+
+                // Process the unfinished work again. This will cause a rehash happening and it should succeed now.
+                boolean needsInputNow = ((SetBuilderOperator) setBuilderOperator).needsInput();
```

Comment:
It might make sense to just put `assertTrue(operator.needsInput())` into `finishOperatorWithYieldingGroupByHash` at the same place where we trigger a rehash; so we can reuse the framework:
```java
// Trigger a process with getOutput() or needsInput()
output = operator.getOutput();
assertTrue(operator.needsInput());
if (output != null) {
    result.add(output);
}

Reply:
ok